go/src/github.com/mesosphere/rendler/scheduler/main.go

```diff
@@ -1,281 +1,284 @@
+	for cpus >= TASK_CPUS && mem >= TASK_MEM {
+		count++
+		cpus -= TASK_CPUS
+		mem -= TASK_MEM
 	}
 
+	return count
+}
+
+type RendlerScheduler struct {
+	tasksCreated int
+	tasksRunning int
+
+	crawlQueue    *list.List
+	renderQueue   *list.List
+	processedURLs *list.List
+	crawlResults  *list.List
+	renderResults map[string]string
+
+	crawlExecutor  *mesos.ExecutorInfo
+	renderExecutor *mesos.ExecutorInfo
+
+	// this channel is closed when the program receives an interrupt,
+	// signalling that the program should shut down.
+	shutdown chan interface{}
+	// this channel is closed when there are no more outstanding tasks,
+	// and that the scheduler does not intend to receive more tasks
+	done chan interface{}
+}
+
+func NewRendlerScheduler(seedUrl string) *RendlerScheduler {
 	// TODO(nnielsen): In local mode, verify artifact locations.
 	rendlerArtifacts := executorURIs()
 
-	crawlExecutor := &mesos.ExecutorInfo{
-		ExecutorId: &mesos.ExecutorID{Value: proto.String("crawl-executor")},
-		Command: &mesos.CommandInfo{
-			Value: proto.String(crawlCommand),
-			Uris:  rendlerArtifacts,
+	s := &RendlerScheduler{
+		crawlQueue:    list.New(),
+		renderQueue:   list.New(),
+		processedURLs: list.New(),
+		crawlResults:  list.New(),
+		renderResults: make(map[string]string),
+
+		crawlExecutor: &mesos.ExecutorInfo{
+			ExecutorId: &mesos.ExecutorID{Value: proto.String("crawl-executor")},
+			Command: &mesos.CommandInfo{
+				Value: proto.String(crawlCommand),
+				Uris:  rendlerArtifacts,
+			},
+			Name: proto.String("Crawler"),
 		},
-		Name: proto.String("Crawler"),
-	}
 
-	renderExecutor := &mesos.ExecutorInfo{
-		ExecutorId: &mesos.ExecutorID{Value: proto.String("render-executor")},
-		Command: &mesos.CommandInfo{
-			Value: proto.String(renderCommand),
-			Uris:  rendlerArtifacts,
+		renderExecutor: &mesos.ExecutorInfo{
+			ExecutorId: &mesos.ExecutorID{Value: proto.String("render-executor")},
+			Command: &mesos.CommandInfo{
+				Value: proto.String(renderCommand),
+				Uris:  rendlerArtifacts,
+			},
+			Name: proto.String("Renderer"),
 		},
-		Name: proto.String("Renderer"),
-	}
 
-	makeTaskPrototype := func(offer mesos.Offer) *mesos.TaskInfo {
-		taskId := tasksCreated
-		tasksCreated++
-		return &mesos.TaskInfo{
-			TaskId: &mesos.TaskID{
-				Value: proto.String(fmt.Sprintf("RENDLER-%d", taskId)),
-			},
-			SlaveId: offer.SlaveId,
-			Resources: []*mesos.Resource{
-				mesos.ScalarResource("cpus", TASK_CPUS),
-				mesos.ScalarResource("mem", TASK_MEM),
-			},
-		}
+		shutdown: make(chan interface{}),
+		done:     make(chan interface{}),
 	}
+	s.crawlQueue.PushBack(seedUrl)
+	return s
+}
 
-	makeCrawlTask := func(url string, offer mesos.Offer) *mesos.TaskInfo {
-		task := makeTaskPrototype(offer)
-		task.Name = proto.String("CRAWL_" + *task.TaskId.Value)
-		task.Executor = crawlExecutor
-		task.Data = []byte(url)
-		return task
+func (s *RendlerScheduler) makeTaskPrototype(offer *mesos.Offer) *mesos.TaskInfo {
+	taskId := s.tasksCreated
```

Comment:
taskID, not taskId. https://code.google.com/p/go-wiki/wiki/CodeReviewComments#Initialisms